src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRazorPageFactoryProvider.cs

```diff
@@ -59,10 +50,23 @@ public RazorPageFactoryResult CreateFactory(string relativePath)
                 // For tilde slash paths, drop the leading ~ to make it work with the underlying IFileProvider.
                 relativePath = relativePath.Substring(1);
             }
-            var result = CompilerCache.GetOrAdd(relativePath, _compileDelegate);
+
+            var result = _razorCompiler.Compile(relativePath);
             if (result.Success)
             {
-                return new RazorPageFactoryResult(result.PageFactory, result.ExpirationTokens, result.IsPrecompiled);
+                var compiledType = result.CompiledType;
+
+                var newExpression = Expression.New(compiledType);
+                var pathProperty = compiledType.GetTypeInfo().GetProperty(nameof(IRazorPage.Path));
+
+                // page.Path = relativePath;
```

Comment:
Yup - it's sort of showing what the expression tree is generating. I can remove it if it's confusing or not obvious.